Use ide login library

chanseokoh opened the pull request:
Fixes #480.

- Uses an external browser
- `GoogleLoginService` creates and maintains one instance of `GoogleLoginState`, which is the core class of the login library
- Made `GoogleLoginService` singleton to do unit testing

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.login/.classpath:
```diff
@@ -4,10 +4,12 @@
 	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
 	<classpathentry kind="src" path="src"/>
 	<classpathentry kind="src" path="generated-sources"/>
+	<classpathentry exported="true" kind="lib" path="lib/com.google.cloud.tools.ide.login-0.0.0-SNAPSHOT.jar"/>
```
I thought this was a bundle already…?  We should be able to bring it in like we do app-tools-lib

chanseokoh replied:
Yeah, this is already a bundle. I couldn't find it in the Maven Central though. I guess someone (us?) need to publish it to the Maven Central first?

elharo replied:
Please file a high-priority issue to do this

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginCommandHandler.java:
```diff
@@ -1,58 +1,44 @@
-    commandService.refreshElements(
-        "com.google.cloud.tools.eclipse.appengine.login.commands.loginCommand", null); //$NON-NLS-1$
-
+    if (credential != null) {
```
I thought you were able to move away from this code?

chanseokoh replied:
It was suggested that we just forget about #454, not putting any further effort to make it work.

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginService.java:
```diff
@@ -16,36 +16,54 @@
-  ));
+  private static final SortedSet<String> OAUTH_SCOPES = Collections.unmodifiableSortedSet(
+      new TreeSet<>(Arrays.asList(
+          "email", //$NON-NLS-1$
```
We need `email` to show their name on the login widgets too, don't we.

chanseokoh replied:
Yeah, eventually we need a fancier UI. The UI... doesn't exist yet. I will create an issue for having an enhanced UI.

patflynn replied:
I believe you need https://www.googleapis.com/auth/appengine.admin for app engine deployment.

chanseokoh replied:
So far, we didn't need it for standard app engine deployment.

patflynn replied:
that is...odd.  Perhaps they've finally sorted this out. I'll check with
the cloud SDK team.

On Tue, Aug 2, 2016 at 1:32 PM, Chanseok Oh <notifications@github.com>
wrote:

> In
> plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginService.java
> <https://github.com/GoogleCloudPlatform/gcloud-eclipse-tools/pull/486#discussion_r73200337>
> :
>
> >
> >    // For the detailed info about each scope, see
> >    // https://github.com/GoogleCloudPlatform/gcloud-eclipse-tools/wiki/Cloud-Tools-for-Eclipse-Technical-Design#oauth-20-scopes-requested
> > -  private static final List<String> OAUTH_SCOPES = Collections.unmodifiableList(Arrays.asList(
> > -      "https://www.googleapis.com/auth/cloud-platform" //$NON-NLS-1$
> > -  ));
> > +  private static final SortedSet<String> OAUTH_SCOPES = Collections.unmodifiableSortedSet(
> > +      new TreeSet<>(Arrays.asList(
> > +          "email", //$NON-NLS-1$
>
> So far, we didn't need it for standard app engine deployment.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/gcloud-eclipse-tools/pull/486/files/9d783a4a3b65898b0709d531ea4d93bfe87480b0#r73200337>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHf5HWpYbjQMRoDvlt55YRW2EGeK4V57ks5qb38wgaJpZM4JakbD>
> .
>

akerekes commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/messages.properties:
```diff
@@ -1,4 +1,6 @@
-LOGIN_BROWSER_TITLE=Sign in to Google
+LOGIN_ERROR_CANNOT_OPEN_BROWSER=Error opening a system browser. Signing in requires an open browser.
+LOGIN_ERROR_DIALOG_MESSAGE=An error occurred while trying to sign in.
+LOGIN_ERROR_DIALOG_TITLE=Signing In Error
```
`Sign in failed`?

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/LoginServiceUi.java:
```diff
@@ -0,0 +1,87 @@
+      return dialog.getValue();
+
+    } catch (PartInitException pie) {
+      Shell shell = PlatformUI.getWorkbench().getDisplay().getActiveShell();
```
Use SWT's `Program.launch(urlString)` — it takes URLs and will use the system's default browser.

chanseokoh replied:
Thanks!

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/LoginServiceUi.java:
```diff
@@ -0,0 +1,87 @@
+import org.eclipse.ui.browser.IWorkbenchBrowserSupport;
+import org.eclipse.ui.commands.ICommandService;
+
+public class LoginServiceUi implements UiFacade {
```
`LoginServiceUi` should have a constructor that takes the relevant `Shell`.  Then you can remove all `PlatformUI` references

chanseokoh replied:
I noticed a slight problem here. There should be a single instance of `GoogleLoginState`, which by its design takes an instance of `UiFacade`. That instance of `UiFacade` lives in the lifetime of `GoogleLoginState`, so clients can't dynamically pass a new instance of `UiFacade` accepting a shell that they see fit. Thoughts?

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/TransientOAuthDataStore.java:
```diff
@@ -0,0 +1,67 @@
+/**
+ * Provides a transient store for saving and loading {@link OAuthData} (a user credential).
+ */
+public class TransientOAuthDataStore implements OAuthDataStore {
```
I think a better approach here is to have the `IEclipseContext` be provided in the constructor and let the creator decide which context should be used.  You could even mark the Constructor with `@Inject` so that it can be created and injected (cf. `ContextInjectionFactory#make(Class,IEclipseContext)`)

akerekes replied:
I agree with @briandealwis I was also looking into this code, but wasn't sure how to solve it. I guess `PlatformUI.getWorkbench().getService(IEclipseContext.class)` should always return the same instance in all reasonable cases?

chanseokoh replied:
I changed the constructor to accept `IEclipseContext`. Didn't use the `@Inject`. (Wasn't sure if you meant injecting `IEcipseContext` into `TransientOAuthDataStore`, or making `TransientOAuthDataStore` injectable by putting it in `EclipseContext`). Maybe next time, after I study the E4 Dependency Injection in depth.

akerekes commented on plugins/com.google.cloud.tools.eclipse.appengine.login.test/META-INF/MANIFEST.MF:
```diff
@@ -8,7 +8,8 @@ Bundle-Vendor: Google, Inc.
 Require-Bundle: org.hamcrest;bundle-version="1.1.0",
- org.junit;bundle-version="4.12.0"
+ org.junit;bundle-version="4.12.0",
+ com.google.cloud.tools.eclipse.test.dependencies
```
Is this required? We import packages below and if something is missing, shouldn't we follow the same pattern to use Import-Package?

chanseokoh replied:
Removed `com.google.cloud.tools.eclipse.test.dependencies` and added ` org.mockito.invocation;provider=google;version="1.10.19"` under `Import-Package`.

akerekes commented on plugins/com.google.cloud.tools.eclipse.appengine.login.test/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginServiceTest.java:
```diff
@@ -0,0 +1,97 @@
+      )));
+
+  @Before
+  public void Setup() {
```
setUp

akerekes commented on plugins/com.google.cloud.tools.eclipse.appengine.login.test/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginServiceTest.java:
```diff
@@ -0,0 +1,97 @@
+  @Mock UiFacade uiFacade;
+  @Mock LoggerFacade loggerFacade;
+
+  GoogleLoginService loginService;
```
Is this used? All methods seem to create their on local instances.

akerekes commented on plugins/com.google.cloud.tools.eclipse.appengine.login.test/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginServiceTest.java:
```diff
@@ -0,0 +1,97 @@
+    Assert.assertNull(loginService.getCachedActiveCredential());
+  }
+
+  OAuthData singleStorageForDataStore;
```
is this used?

chanseokoh replied:
Good catch.

akerekes commented on plugins/com.google.cloud.tools.eclipse.appengine.login.test/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginServiceTest.java:
```diff
@@ -0,0 +1,97 @@
+  }
+
+  @Test
+  public void testClearSavedCredentialIfNullRefreshToken() {
```
use `_` to separate method name and test case

akerekes commented on plugins/com.google.cloud.tools.eclipse.appengine.login.test/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginServiceTest.java:
```diff
@@ -0,0 +1,97 @@
+  }
+
+  @Test
+  public void testClearSavedCredentialIfScopesChanged() {
```
use `_` to separate method name and test case

akerekes commented on plugins/com.google.cloud.tools.eclipse.appengine.login.test/src/com/google/cloud/tools/eclipse/appengine/login/TransientOAuthDataStoreTest.java:
```diff
@@ -0,0 +1,77 @@
+        singleStorageForIEclipseContext = invocation.getArgumentAt(1, OAuthData.class);
+        return null;
+      }
+    }).doThrow(Exception.class).when(eclipseContext).set(anyString(), any());
```
Is the `doThrow()` invocation used in this test?

chanseokoh replied:
It is to fail if `set()` is called twice. (If called twice, this test setup won't work as intended.)

akerekes commented on plugins/com.google.cloud.tools.eclipse.appengine.login.test/src/com/google/cloud/tools/eclipse/appengine/login/TransientOAuthDataStoreTest.java:
```diff
@@ -0,0 +1,77 @@
+  OAuthData singleStorageForIEclipseContext;
+
+  @Test
+  public void testSaveAndLoadOAuthData() {
```
What's the goal of this test? If it's to ensure that `loadAuthData()` calls `eclipseContext.get()` then we can verify that without the `answer` setup.

chanseokoh replied:
It's a test that `loadAuthData()` returns the `OAuthData` object set through `saveOAuthData()`. It took me some thing to make this work with Mockito. I couldn't find a better than the `Answer` setup.

akerekes commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginService.java:
```diff
@@ -54,21 +72,16 @@
-      IEclipseContext eclipseContext = PlatformUI.getWorkbench().getService(IEclipseContext.class);
-      eclipseContext.set(STASH_OAUTH_CRED_KEY, credential);
+  public Credential getActiveCredential() {
+    // TODO: holding a lock for a long period of time (especially when waiting for UI events)
```
I'm a bit worried that we'll lock up here some action and the user will have no feedback on what's going on.
How about using timeouts when trying to get the lock and throw and exception to notify other threads trying to login, that one's already in progress?

chanseokoh replied:
Good idea, but I think that'd be of no use and in fact not necessary, because this method is called only from the UI-thread. The UI-thread will never be blocked. (I made it that other methods that can be called from non-UI threads and holds a conflicting lock are always terminating quickly.) It's just that, if the UI-thread is holding a lock for a long time for taking care of login, other headless-jobs are blocked. Ideally, I think we should fix the ide-login library itself.

chanseokoh replied:
On second thought, I wonder if we have two workbenches and both try to login. If two workbenches share a single class loader (i.e., share a single `GoogleLoginService` instance), then I think there is a problem. Not sure how I can actually test this.

chanseokoh replied:
Seems like workbenches are per-JVM basis, so I think we are covered and okay.

briandealwis replied:
Not in a RAP situation.  We don't have that as a requirement, but it's nicer not to make assumptions (and better code) — I'll add a bit about this separately.

chanseokoh replied:
In that case, here's my idea: check if the login is currently in process (I think a `volatile boolean` flag would work) in `getActiveCredential()`, and if it is, it shows an error message dialog saying "Login already in progress. Please retry later."

akerekes replied:
That's why I was thinking about locks. If you use the flag, and it's false (meaning no login is in progress), by the time you get to the next line and start the synchronized block, it can be changed to true by a concurrent thread. Possibly a very small chance, but it's still there.

briandealwis replied:
Sounds like we want to be able to chain a set of actions together.  Login could be done in a UIJob, hands off to a XXXJob, etc.  With some ability to handle an OK status, cancelled status (dialog: deploy cancelled), or error status (error dialog).

chanseokoh replied:
@akerekes I believe atomic-get-and-set should work. If it is set, error and return. If it is not set, set it and grab a lock.

chanseokoh replied:
@briandealwis Yeah, that'd be cool. The problem here is that the entire stream of login work is done by the ide-login library we can't control and the library mixes UI flow and non-UI flow in itself.

briandealwis commented:
@chanseok I'm not particularly enamoured by turning GoogleLoginService into a singleton — I've just seen far too many problems with singletons.  I wondered what you thought about instead changing it to be an OSGi Service, and plus we can replace the instance for testing by defining a higher ranking (`org.osgi.framework.Constants#SERVICE_RANKING`).  The consumers should be able to use the `IServiceLocator` interface, implemented by most workbench components, to obtain the service, but we can turn `getInstance()` into a helper `getInstance(Class)` that obtains the service from the class' bundle if necessary.

chanseokoh commented:
@briandealwis I see. Let's avoid making a singleton. Finally, it's time for me to get to know about creating an OSGi service!

briandealwis commented:
The easiest way to do it is via Declarative Services, which uses an XML file to define the service.  Take a look at the `gcloud.xml` in the `.sdk` bundle, and how it's referenced from the `MANIFEST.MF`.  Your bundle has to be marked with 'activate on class load' (= `Bundle-ActivationPolicy: lazy` in the `MANIFEST.MF`).

chanseokoh commented:
@briandealwis wow, thanks!

chanseokoh commented:
I'm still working on making it an OSGi service.

chanseokoh commented:
I made it an OSGi service. I think the way it is registered and retrieved is a bit different from `.sdk`, but it's working anyway.

- Defined `IGoogleLoginService`.
- Added `Activator`, which registers a service instance.
- Added `OGSI-INF/login.xml` and a corresponding entry in `MANIFEST.MF`.
- Consumers call `PlatformUI.getWorkbench().getService(IGoogleLoginService.class)` to get the service instance.

I assume here that the lifetime of a service instance aligns with the lifetime of Eclipse (or workbench), essentially making it a singleton. If it isn't the case, we'll be in slight trouble.

One warning message I see when starting up:

```
!ENTRY com.google.cloud.tools.eclipse.appengine.login 4 0 2016-08-03 11:54:24.234
!MESSAGE [SCR] Component definition XMLs not found in bundle com.google.cloud.tools.eclipse.appengine.login. The component header value is OGSI-INF/login.xml
```

I don't understand this, since the file is there and everything is working.

akerekes commented on plugins/com.google.cloud.tools.eclipse.appengine.login/META-INF/MANIFEST.MF:
```diff
@@ -6,20 +6,21 @@ Bundle-Version: 0.1.0.qualifier
 Bundle-ActivationPolicy: lazy
-Require-Bundle: com.google.cloud.tools.app.lib;bundle-version="0.1.0",
- com.google.gson,
+Service-Component: OGSI-INF/login.xml
```
Should be OSGI-INF/...

chanseokoh commented:
I realized I was doing the OGSi service stuff in a convoluted way. The last commit removes the `Activator`. You can start taking a look.

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployCommandHandler.java:
```diff
@@ -88,8 +86,10 @@ private IPath createWorkDirectory() throws IOException {
-    Credential credential = new GoogleLoginService().getActiveCredential(shellProvider);
+  private Credential login() throws CoreException {
+    IGoogleLoginService loginService =
+        PlatformUI.getWorkbench().getService(IGoogleLoginService.class);
```
We want to avoid using `PlatformUI.getWorkbench()` wherever possible.  In this case, we're provided the `ExecutionEvent` which provides access to a bunch of objects that implement `IServiceLocator`.  Here's what I'd do:
  - pass in the `IGoogleLoginService` instance into `login()`
  - create a `getService(ExecutionEvent,Class<?>)` method that calls the following sequence of methods to find an `IServiceLocator`:
     - `HandlerUtil.getActiveSite()` (incl views and editors)
     - `HandlerUtil.getActiveEditor()`
     - `HandlerUtil.getActiveWorkbenchWindow()`
     - and if none of them work, then `PlatformUI.getWorkbench()`

nbashirbello replied:
Maybe something like:

GoogleLoginServer.getInstance() returning IGoogleLoginService 

public static IGoogleLoginService getInstance() {
  if (instance == null) {
    instance = new GoogleLoginService();
  }
  return instance;
}

chanseokoh replied:
@nbashirbello It was like that before, but it was suggested we avoid doing a singleton and rather make this an OSGi service.

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.login/build.properties:
```diff
@@ -6,6 +6,7 @@ bin.includes = META-INF/,\
                icons/,\
-               lib/
+               lib/,
+               OSGI-INF/login.xml
```
Could just make this `OSGI-INF/` in case any other services are added

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginCommandHandler.java:
```diff
@@ -19,40 +15,33 @@
 
 public class GoogleLoginCommandHandler extends AbstractHandler implements IElementUpdater {
 
+  private IGoogleLoginService getGoogleLoginService() {
```
Ditto about avoiding _PlatformUI.getWorkbench()_ made in [_StandardDeployCommandHandler_](https://github.com/GoogleCloudPlatform/gcloud-eclipse-tools/pull/486/files#r73461563)

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginService.java:
```diff
@@ -16,96 +16,108 @@
-    return (Credential) eclipseContext.get(STASH_OAUTH_CRED_KEY);
+  public GoogleLoginService() {
+    this(new TransientOAuthDataStore(PlatformUI.getWorkbench().getService(IEclipseContext.class)),
+        new LoginServiceUi(), new LoginServiceLogger());
```
This is where I'd pass in the _Display_ or _Shell_ to `LoginServiceUi`

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginService.java:
```diff
@@ -16,96 +16,108 @@
+  @Override
+  public Credential getActiveCredential() {
+    if (!loginInProgress.compareAndSet(false, true)) {
+      LoginServiceUi.showErrorDialogHelper(
```
Why explicit use of LoginServiceUi instead of the uiFacade passed into the `GoogleLoginState`

chanseokoh replied:
This method is static, so using an instance instead of calling statically actually generates a warning. You'are saying something else?

briandealwis replied:
I see now the fully parameterized `GoogleLoginServer` constructor is `@VisibleForTesting`.  Was thinking that since much of the UI interaction has been made pluggable with the `UiFacade`, we may as well use it here too.

chanseokoh replied:
The UI interaction is not flexibly pluggable unfortunately. `GoogleLoginState` should be constructed only once as a singleton (it maintains a login-related state) and the `UiFacade` is fixed at initialization.

chanseokoh replied:
BTW, `showErrorDialogHelper` is actually just a helper method to display an error dialog. Probably it should go into the `.ui.util` bundle.

briandealwis replied:
Hmm, I don't see how it's more helpful than just calling `MessageDialog.openError(shell, title, message)`.  Regardless of the pluggability of the UiFacade, you have a UiFacade and I don't understand why you wouldn't just use its `UiFacade#showErrorDialog(title,message)`.

chanseokoh replied:
No problem. I'll use `UiFacade#showErrorDialog(title,message)`.

chanseokoh replied:
Ah, the problem was that, because of [this](https://github.com/GoogleCloudPlatform/gcloud-eclipse-tools/pull/486/files/535f3b5b0506cf3fab62fd9015202df008d4e085#r73581984), `title` and `message` are completely ignored in `UiFacade#showErrorDialog(title,message)`. The method is not usable.

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginService.java:
```diff
@@ -16,96 +16,108 @@
-        GoogleLoginBrowser.REDIRECT_URI);
-
-    return createCredential(authRequest.execute());
+    // TODO: holding a lock for a long period of time (especially when waiting for UI events)
```
Should this method return a Future?

chanseokoh replied:
Are you suggesting making `getActiveCredential()` return a Future or making this portion use a Future? If the latter case, that won't really help anything, because (if I'm not mistaken) this method still needs to block on `Future.get()` while holding a lock while the Future computation (i.e., login through UI) is on-going.

chanseokoh replied:
Ah, sorry. You explicitly said returning a Future. I'll go with that.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login.test/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginServiceTest.java:
```diff
@@ -0,0 +1,87 @@
+@RunWith(MockitoJUnitRunner.class)
+public class GoogleLoginServiceTest {
+
+  @Mock OAuthDataStore dataStore;
```
nit: these fields can probably be private

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login.test/src/com/google/cloud/tools/eclipse/appengine/login/TransientOAuthDataStoreTest.java:
```diff
@@ -0,0 +1,62 @@
+@RunWith(MockitoJUnitRunner.class)
+public class TransientOAuthDataStoreTest {
+
+  @Mock IEclipseContext eclipseContext;
```
private

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login/META-INF/MANIFEST.MF:
```diff
@@ -7,19 +7,19 @@ Bundle-Vendor: Google, Inc.
+ lib/google-http-client-jackson-1.22.0.jar,
  lib/google-oauth-client-1.22.0.jar,
- lib/jackson-core-2.1.3.jar,
+ lib/jackson-core-asl-1.9.13.jar,
```
I'm surprised we can't get this one from maven

elharo commented:
The non-google jackson jars should be pulled form maven: https://mvnrepository.com/artifact/com.fasterxml.jackson.core/jackson-core/2.1.3

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/LoginServiceUi.java:
```diff
@@ -0,0 +1,85 @@
+  }
+
+  @Override
+  // TODO(chanseok): the login library doesn't call this method for logging in. Fix the library.
```
This would be better as an issue rather than a TODO, since no change to this code is needed here. Please file one.

chanseokoh replied:
My mistake. I looked at the ide-login code again, and it ends up calling this method. Problem solved!

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/LoginServiceUi.java:
```diff
@@ -0,0 +1,85 @@
+      return null;
+    }
+
+    // TODO(chanseok): remove InputDialog and set up a web server to receive authorization code.
```
This one should probable be an issue and a TODO. Please file an enhancement issue that would be blocked on this PR.

chanseokoh replied:
Created #503.

chanseokoh commented:
I'll try pulling the jackson jars from the Maven Central again. (I failed it last time.)

nbashirbello commented on plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployCommandHandler.java:
```diff
@@ -88,8 +86,10 @@ private IPath createWorkDirectory() throws IOException {
+  private Credential login() throws CoreException {
+    IGoogleLoginService loginService =
+        PlatformUI.getWorkbench().getService(IGoogleLoginService.class);
+    Credential credential = loginService.getActiveCredential();
```
I think thee GoogleLoginService should probably have and isLoggedIn() method and that can be called here.

chanseokoh replied:
@akerekes and I discussed this before: #472

nbashirbello commented on plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployCommandHandler.java:
```diff
@@ -1,7 +1,14 @@
 package com.google.cloud.tools.eclipse.appengine.deploy.standard;
```
License header?

nbashirbello commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/TransientOAuthDataStore.java:
```diff
@@ -0,0 +1,54 @@
+
+/**
+ * Provides a transient store for saving and loading {@link OAuthData} (a user credential).
+ */
```
Just to confirm, this is a temporary solution to storing the credentials right?

Does this support Eclipse restarting?

chanseokoh replied:
You're right. It's almost like providing an empty implementation to the `OAuthDataStore`.

chanseokoh commented:
This PR fixes #446 too.

nbashirbello commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/LoginServiceUi.java:
```diff
@@ -0,0 +1,85 @@
+  }
+
+  @Override
+  public void showErrorDialog(String title, String message) {
```
The title and message are supposed to be passed by the feature trying to log in. For instance, if a user calls deploy without logging in, deploy would make an attempt to log in. If there is an error and we pop up a dialog, the title should say "Deploying project ...", so that the user is clear why logging was called.

chanseokoh replied:
I agree. That should be the case, but the ide-login library is [broken in this regard](https://github.com/GoogleCloudPlatform/ide-login/blob/master/src/main/java/com/google/cloud/tools/ide/login/GoogleLoginState.java#L297). Both the title and the message are hard-coded (not localized either).

I created an issue against the login library: https://github.com/GoogleCloudPlatform/ide-login/issues/16

elharo replied:
Are you sure about this? I thought login was the result of an explicit user selection of a login button/menu item. What am I missing here?

chanseokoh replied:
We may enforce the policy that login is done only through explicit user selection of a button or menu. Currently, the deployment automatically triggers login rather than simply failing with a message saying that the user needs to login first. Here is the relevant issue: #472. We could make a decision soon.

chanseokoh commented:
Now pulling `jackson-core-asl-1.9.13.jar` from the Maven Central. Builds fine locally, and the login is working properly. PTAL.

elharo commented:
Can you also remove plugins/com.google.cloud.tools.eclipse.appengine.login/lib/jackson-core-2.1.3.jar? It looks like you're not actually referencing it.

chanseokoh commented:
@elharo The jar is being removed as part of this PR. (You can find it in the PR if you look really hard.)

elharo commented:
LGTM

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/LoginServiceUi.java:
```diff
@@ -0,0 +1,86 @@
+    display.asyncExec(new Runnable() {
+      @Override
+      public void run() {
+        PlatformUI.getWorkbench().getService(ICommandService.class).refreshElements(
```
Here you're using `IWorkbench` as an IServiceLocator.  Rather than hard-code the workbench here, pass in `IServiceLocator` in the constructor.

briandealwis commented on plugins/com.google.cloud.tools.eclipse.util/src/com/google/cloud/tools/eclipse/util/ServiceUtils.java:
```diff
@@ -0,0 +1,55 @@
+
+public class ServiceUtils {
+
+  public static <T> T getService(ExecutionEvent event, Class<T> api) {
```
We should have a javadoc here

chanseokoh commented:
Had a chat with @briandealwis and learned that intrinsic references/access to workbenches is bad.

The last commit does the following:

  - `LoginServiceUi` is now free of any workbench, accepting `IServiceLocator` and `IShellProvider`.
  - `GoogleLoginService` is initialized by `activate()` (which is called by OSGi DS). Before, it was done by a 0-arg constructor. @briandealwis, does this look good? (I hope it is not initialized multiple times for an instance.)
  - Added javadoc to `ServiceUtils`.

briandealwis commented:
LGTM.  But if you want to keep `GoogleLoginService` as a POJO, you could:
  - drop the `ComponentContext` argument on `activate`: it's optional, and only required to access metadata set in the `OSGI-INF/XXX.xml` file or to access the OSGi framework.
  - keep the 3-arg constructor or add setters for the datastore, UI facade, and logging facade
  - merge your `activate` and `initialize` methods to set defaults for the datastore, UI and logging facades as needed.

Then your OSGI-DS created service will be no different from running:
```
   GoogleLoginService service = new GoogleLoginService();
   service.activate();
```
OSGi DS is then just an activation and service access mechanism.

chanseokoh commented:
Removed the `ComponentContext` argument. Thanks.

> - keep the 3-arg constructor or add setters for the datastore, UI facade, and logging facade
> - merge your activate and initialize methods to set defaults for the datastore, UI and logging facades as needed.

While I was trying these, I encountered some clunkiness. Setters don't make much sense, as these three arguments are basically meant for creating `GoogleLoginState`. `initialize()` (and the 3-arg constructor) does some extra setup too. So, to reduce code duplication, you anyway end up defining a common init method if you want to keep a constructor. The constructor was only for unit testing, so I think it's okay if we don't have it. The reason for a separate `initialize()` is also for unit testing.

Anyway, without the last two bullets, we can still achieve

```
   GoogleLoginService service = new GoogleLoginService();
   service.activate();
```

briandealwis commented:
LGTM

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginService.java:
```diff
@@ -60,7 +60,7 @@
+  protected void activate() {
     IWorkbench workbench = PlatformUI.getWorkbench();
     IEclipseContext eclipseContext = workbench.getService(IEclipseContext.class);
     IShellProvider shellProvider = new SameShellProvider(workbench.getDisplay().getActiveShell());
```
Oops, that's not what you want: that will fix the service with the shell of the active window / dialog _at creation time_.  I think you want:
```
IShellProvider shellProvider = new IShellProvider() {
  public Shell getShell() {
    return workbench.getDisplay().getActiveShell();
  }
}
```

briandealwis commented:
LEBTM (looks even better…)

briandealwis commented:
LGTM
